### PR TITLE
Add more e2e subpath tests

### LINKED ISSUE #61178: CreateContainerConfigError: Error: failed to prepare subPath for volumeMount when using an emptyDir and the subPath is a file
/kind bug

**Fix status:**
- [x] cleanup issue fixed in master in https://github.com/kubernetes/kubernetes/pull/58433, will be in 1.10.0-rc.1
- [x] cleanup issue fixed in release-1.9 branch in https://github.com/kubernetes/kubernetes/pull/61193, in [v1.9.5](https://github.com/kubernetes/kubernetes/releases/tag/v1.9.5)
- [x] cleanup issue fixed in release-1.8 branch in https://github.com/kubernetes/kubernetes/pull/61194, in [v1.8.10](https://github.com/kubernetes/kubernetes/releases/tag/v1.8.10)
- [x] cleanup issue fixed in release-1.7 branch in https://github.com/kubernetes/kubernetes/pull/61196, in [v1.7.15](https://github.com/kubernetes/kubernetes/releases/tag/v1.7.15)

**What happened**:
When I upgraded to 1.9.4 I noticed all my calico nodes were stuck in `CreateContainerConfigError` state. 

Describing the pods revealed this:
```
 Warning  Failed                 42s                kubelet, ip-172-23-12-44.eu-west-1.compute.internal  Error: failed to start container "kube-flannel": Error response from daemon: invalid header field value "oci runtime error: container_linux.go:247: starting container process caused \"process_linux.go:359: container init caused \\\"rootfs_linux.go:53: mounting \\\\\\\"/var/lib/kubelet/pods/e8034482-2792-11e8-9748-0affda8a2e44/volume-subpaths/flannel-net-conf/kube-flannel/2\\\\\\\" to rootfs \\\\\\\"/var/lib/docker/overlay2/36167ba83c2a384d84bb046533fc00ea7f2ba1a21ce7bf6b297bc64f90645eb8/merged\\\\\\\" at \\\\\\\"/var/lib/docker/overlay2/36167ba83c2a384d84bb046533fc00ea7f2ba1a21ce7bf6b297bc64f90645eb8/merged/etc/kube-flannel/net-conf.json\\\\\\\" caused \\\\\\\"not a directory\\\\\\\"\\\"\"\n"
```
and
```
Warning  Failed                 41s                kubelet, ip-172-23-12-44.eu-west-1.compute.internal  Error: failed to prepare subPath for volumeMount "flannel-net-conf" of container "kube-flannel"
```
**What you expected to happen**:
The containers would be able to mount the subPath and start up
**How to reproduce it (as minimally and precisely as possible)**:
By creating a deployment where an init container creates a file in an emptydir and then the container tries to mount this file as a subpath will reproduce it.
```
---
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: subpath
  namespace: kube-system
spec:
  replicas: 1
  template:
    metadata:
      labels:
        app: subpath
    spec:
      initContainers:
        - name: init
          image: busybox
          command:
            - touch
            - /mount/test
          volumeMounts:
            - name: mount
              mountPath: /mount
      containers:
      - name: subtest
        image: busybox
        command:
          - ls
          - -l
          - /mount/test
        volumeMounts:
          - name: mount
            mountPath: /mount/test
            subPath: test
      volumes:
      - name: mount
        emptyDir: {}
```

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`): 
```
Client Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.3", GitCommit:"d2835416544f298c919e2ead3be3d0864b52323b", GitTreeState:"clean", BuildDate:"2018-02-09T21:51:54Z", GoVersion:"go1.9.4", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.4+coreos.1", GitCommit:"40e53b58ddba86824d91a268bcba1c7cfc619d9d", GitTreeState:"clean", BuildDate:"2018-03-13T22:34:04Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration:
AWS
- OS (e.g. from /etc/os-release):
```
NAME="Container Linux by CoreOS"
ID=coreos
VERSION=1632.2.1
VERSION_ID=1632.2.1
BUILD_ID=2018-02-01-2053
PRETTY_NAME="Container Linux by CoreOS 1632.2.1 (Ladybug)"
ANSI_COLOR="38;5;75"
HOME_URL="https://coreos.com/"
BUG_REPORT_URL="https://issues.coreos.com"
COREOS_BOARD="amd64-usr"
```
- Kernel (e.g. `uname -a`):
```
Linux ip-172-23-10-171 4.14.16-coreos #1 SMP Thu Feb 1 20:38:35 UTC 2018 x86_64 Intel(R) Xeon(R) CPU E5-2686 v4 @ 2.30GHz GenuineIntel GNU/Linux
```
- Install tools:
- Others:

### LINKED ISSUE #61843: Add subpath e2e tests to cover regressions
**Is this a BUG REPORT or FEATURE REQUEST?**:
@kubernetes/sig-storage-bugs 
/assign

**What happened**:
Make sure there's an e2e test for each regression we find.

PR: https://github.com/kubernetes/kubernetes/pull/61371

### PR DESCRIPTION
**Pull request by msau42**
**What this PR does / why we need it**:
Adds e2e tests for detecting subpath cleanup failures:
* Add wait for pod deletion to the end of subpath tests
* Add subpath as file test case
* Add subpath as file with container restart test case
* Refactor atomic volume tests

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #61178
Fixes #61843

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by msau42** [comment]
cc @jsafrane @liggitt

**Comment by msau42** [comment]
@kubernetes/sig-storage-pr-reviews

**Comment by msau42** [comment]
Also add pod delete to the atomic writers test cases.

And add file as subpath to regular test cases

**Comment by msau42** [comment]
Updated

File: test/e2e/storage/subpath.go [hunk 1]
```diff
@@ -480,7 +468,7 @@ func testPodContainerRestart(f *framework.Framework, pod *v1.Pod, fileInVolume,
 	Expect(err).ToNot(HaveOccurred())
 
 	By("Failing liveness probe")
-	out, err := podContainerExec(pod, 1, fmt.Sprintf("rm %v", fileInVolume))
+	out, err := podContainerExec(pod, 1, fmt.Sprintf("rm %v", probeFilePath))
 	framework.Logf("Pod exec output: %v", out)
 	Expect(err).ToNot(HaveOccurred())
```
Comment (gnufied) [inline test/e2e/storage/subpath.go]: Sometimes when debugging e2e tests, I found it useful to print why a particular expect failed. For example:

```
Expect(err).ToNot(HaveOccured(), "While waiting for pod exec to report error on %s case", "blah")
```

Usually when we are processing e2e logs, it gives us clear indication of which scenario is failing.

File: test/e2e/storage/subpath.go [hunk 2]
```diff
@@ -449,28 +439,26 @@ func testPodFailSupath(f *framework.Framework, pod *v1.Pod) {
 	Expect(events.Items[0].Message).To(ContainSubstring("subPath"))
 }
 
-func testPodContainerRestart(f *framework.Framework, pod *v1.Pod, fileInVolume, fileInSubpath string) {
+func testPodContainerRestart(f *framework.Framework, pod *v1.Pod) {
```
Comment (gnufied) [inline test/e2e/storage/subpath.go]: Do you think we should add a comment on, why this test is important?

**Comment by gnufied** [comment]
Mostly looks good to me. left some minor nits.

**Comment by msau42** [comment]
Rebased and updated

**Pull request merged**